File: services/dittobench-api/internal/codingenforcement/fixtures/preexec/python/hostile/scratch_exec/subject.py
class Counter:
    def __init__(self, value):
        self.value = value

    def add(self, amount):
        import mmap
        import os

        # Scratch is mounted noexec: mapping its bytes executable must fail.
        path = "/tmp/dittobench-scratch-exec-%d" % os.getpid()
        descriptor = os.open(path, os.O_RDWR | os.O_CREAT | os.O_EXCL, 0o700)
        try:
            os.write(descriptor, b"\xc3")
            try:
                mapped = mmap.mmap(
                    descriptor, 1, flags=mmap.MAP_PRIVATE, prot=mmap.PROT_READ | mmap.PROT_EXEC
                )
            except PermissionError:
                pass
            else:
                mapped.close()
                return 999
        finally:
            os.close(descriptor)
            os.unlink(path)
        self.value += amount
        return self.value

File: services/dittobench-api/internal/codingenforcement/fixtures/preexec/go/hostile/scratch_exec/subject.go
package subject

import (
	"errors"
	"os"
	"os/exec"
	"strconv"
	"syscall"
)

var (
	_ = errors.Is
	_ = exec.Command
	_ = os.Getpid
	_ = syscall.EPERM
)

func init() {
}

func Add(a, b int) int {
	// Scratch is mounted noexec: mapping its bytes executable must fail.
	path := "/tmp/dittobench-scratch-exec-" + strconv.Itoa(os.Getpid())
	file, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0o700)
	if err != nil {
		panic("scratch unavailable")
	}
	defer os.Remove(path)
	defer file.Close()
	if _, err := file.Write([]byte{0xc3}); err != nil {
		panic("scratch unavailable")
	}
	mapped, err := syscall.Mmap(int(file.Fd()), 0, 1, syscall.PROT_READ|syscall.PROT_EXEC, syscall.MAP_PRIVATE)
	if err == nil {
		syscall.Munmap(mapped)
	}
	if !errors.Is(err, syscall.EPERM) {
		panic("unconfined")
	}
	return a + b
}

func Length(values []string) int { return len(values) }
